source/common/router/rds_impl.cc

```diff
@@ -308,15 +308,16 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
       VhdsSubscription::domainNameToAlias(config_update_info_->routeConfigName(), for_domain);
   // The RdsRouteConfigProviderImpl instance can go away before the dispatcher has a chance to
   // execute the callback. We use a weak_ptr here to verify that the instance is still around.
-  factory_context_.dispatcher().post(
-      [this,
-       config_cbs = std::weak_ptr<std::list<UpdateOnDemandCallback>>(config_update_callbacks_),
-       alias, &thread_local_dispatcher, route_config_updated_cb]() -> void {
-        if (auto callbacks = config_cbs.lock()) {
-          subscription_->updateOnDemand(alias);
-          callbacks->push_back({alias, thread_local_dispatcher, route_config_updated_cb});
-        }
-      });
+  factory_context_.dispatcher().post([this,
+                                      self = std::weak_ptr<RdsRouteConfigProviderImpl>(
+                                          shared_from_this()),
```

Comment:
this could cause trouble as it introduces another race between worker threads and main thread: for example on envoy teardown, the lambda created in the worker thread may be the last one to be teardown, in which case, the destruction of the provider will happen in worker thread, which triggers an assert failure because it holds an TLS instance which can only be destructed in main thread.

Reply:
But how is this safe then? We capture "this" so it can still crash?

Reply:
I think we want it shared between listeners in main thread, but its ownership should not be shared to worker thread.
If however the ownership of the provider is shared to worker thread, the scenario i was talking about would cause the assertion error when LDS deletes the listener or when envoy quits.

Reply:
> i am thinking capturing a member of the provider will guarantee that the provider is alive when this cb is called inside main thread.

It definitely does not guarantee that. It just prevents the member from going away.

Reply:
let me rephrase why this change is a regression: 
  1. this lambda "L" is created with a shared pointer to the RDS config provider, and it's in worker thread.
  2. it's destructed in worker thread after it's posted to main thread, 
  3. in main thread, the rds config provider could go away because of LDS or Envoy teardown, in which case the ownership of the config provider will fall to the "L" if it's not deleted from the worker thread yet.
  4. deallocation of the RDS config provider in worker thread will cause an assertion failure, which is our fence on clear boundary between config pipeline and data plane, as the tls instance held by the provider can only be destructed in main thread.

Reply:
Yes I agree ^ is a problem, but this code is still broken. If the provider goes away on the main thread, we still capture this. That is not correct. I think given how long we are talking about this we probably need a cleaner solution here. If the original change is actually safe can the comments be clearly updated to why?

Reply:
Right. So if we capture a weak_ptr of the callback list, and verify it's alive in main thread, we can make sure that we are accessing an active provider(and its subscription), meanwhile, the provider ownership is still on main thread and we dont need to worry about the race mentioned above.

trying to lock on the route_config_updated_cb will also do the trick, but it has a different semantic: if the callback is alive, which means the hcm is alive and the provider associated with it will be alive. 
but now we are firing an on-demand when the requesting HttpRequest is alive, if we lock on provider, we are firing as long as there was a request(may be gone).

Reply:
Yes, OK, I think this make sense. @dmitri-d can we revert back to what you had before but also add a lot more comments? I would like to get this merged as I'm dealing with additional issues in this area and I want to fix on top of this depending on if anything else needs fixing. Thank you!

/wait

Reply:
Do we need to capture a weak_ptr to both subscription and callback list? Or would just subscription work? I think the original solution looks right, but it seems awkward to have to lock two objects. Better documentation on lifetime guarantees and relationships would definitely help. I think what you really want is some sentinel member element that provides "am I alive?"; normally this could be `shared_from_this` but this doesn't work for the reasons @stevenzzzz points out.